keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -0,0 +1,625 @@
+Note that namespace objects must match both object selector (if specified) and namespace selector to be sent to the Webhook. 
+
+The object selector applies to an Object's labels. For create and update, the object is the new object. For delete, it is existing object (passed as `oldObject` in the `admissionRequest`). For sub-resources, if the subresource accepts the whole object and apply the changes to object's labels (e.g. `pods/status`) the labels on the new object will be used otherwise (e.g. `pods/proxy`) the labels on the existing parent object will be used. If subresource does not have a parent object, the Webhook call may fail or skipped based on the failure policy.
+
+The propose change is to add an ObjectSelector to the webhook API both in v1 and v1beta1:
+
+```golang
+type Webhook struct {
+    ...
+     // ObjectSelector decides whether to run the webhook on an object based
+     // on whether the object.metadata.labels matches the selector. A namespace object must
+     // match both namespaceSelector (if present) and objectSelector (if present).
+     // For sub-resources, if the labels on the newObject is effective (e.g. pods/status)
+     // they will be used, otherwise the labels on the parent object will be used. If the
+     // sub-resource does not have a parent object, the call may fail or skipped based on
+     // the failure policy.
+     //
+     // Default to the empty LabelSelector, which matches everything.
+     // +optional
+     ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,7,opt,name=objectSelector"`
+}
+```
+
+### Scope
+
+Current webhook Rules applies to objects of all scopes. That means a Rule can uses wildcards to target both namespaced and cluster scoped objects. The proposal is to add a scope field to Admission Webhook configuration to limit webhook target on namespaced object or cluster scoped objects. This enables webhook developers to target all namespace objects or all cluster-scoped objects. The field will be added to both v1 and v1beta1. The field is optional and empty value means no scope restriction.
+
+```golang
+type ScopeType string
+
+const (
+     // ClusterScope means that scope is limited to cluster objects.
+     ClusterScope ScopeType = "Cluster"
+     // NamespacedScope means that scope is limited to namespaced objects.
+     NamespacedScope ScopeType = "Namespaced"
+)
+
+type Rule struct {
+    ...
+
+     // Scope specifies the scope of this rule. If unespecified, the scope is
```

Comment:
api question: why not force an explicit all and setting a default?  this is effectively putting defaults in the hands of clients.

Reply:
> this is effectively putting defaults in the hands of clients.

by clients, do you mean webhook admission evaluators?

Reply:
I have a slight aesthetic preference for letting unspecified mean all, rather than defaulting it to `*` (no scope specified means unscoped, which is sort of coherent). That way a fully qualified group/version/resource tuple rule (which is unambiguous) doesn't get useless fields added to it automatically. 

Do you foresee problems with that approach?

Reply:
> by clients, do you mean webhook admission evaluators?

Webhook admission *config* evaluators.  Yes, they are clients that are all agreeing on a default instead of specifying it.  It's easy for golang servers, less easy for others.  I don't like distributed defaulting.  Single actor/consumer defaulting, sure.  Multi actor/consumer defaulting, not so much.